errors.go

```diff
@@ -191,6 +191,24 @@ func errCode(prefix, code string) string {
 	return strings.Join([]string{prefix, code}, ".")
 }
 
+// addParams returns a new error with new params merged into the original error's
+func addParams(err *Error, params map[string]string) *Error {
+	copiedParams := make(map[string]string, len(err.Params)+len(params))
+	for k, v := range err.Params {
+		copiedParams[k] = v
+	}
+	for k, v := range params {
+		copiedParams[k] = v
+	}
+
+	return &Error{
+		Code:        err.Code,
+		Message:     err.Message,
+		Params:      copiedParams,
+		StackFrames: err.StackFrames,
```

Comment:
This is the only bit I'm not sure of. The `Stack` is a `[]*Frame`. This opens up the posibility that a `Frame` in the `Stack` is mutated somewhere else. Looking at the code, such mutation doesn't happen, but I thought I'd mention it.

Reply:
@danielchatfield what do you think about this point?

Reply:
I think it's fine.